File: Question 0103 - Binary Tree Zigzag Level Order Traversal/src/Solution.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();

        if(root == null) return result;

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        boolean leftToRight = true;

        while(!q.isEmpty())
        {
            int size = q.size();
            List<Integer> ans = new ArrayList<>(size);

            //level process
            for (int i = 0; i < size; i++) {
                TreeNode front = q.poll();

                if(front != null)
                {
                    //normal insert or reverse insert
                    if(leftToRight) ans.add(front.val);
                    else ans.add(0,front.val);

                    if(front.left != null) q.add(front.left);
                    if(front.right != null) q.add(front.right);
                }
            }

            //direction change
            result.add(ans);
            leftToRight = !leftToRight;
        }
        return result;
    }
}

File: Question 0103 - Binary Tree Zigzag Level Order Traversal/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode *root)
    {
        vector<vector<int>> ans;
        if (!root)
            return ans;

        queue<TreeNode *> q;
        q.push(root);
        bool leftToRight = true;

        while (!q.empty())
        {
            int sz = q.size();
            vector<int> curr;
            for (int i = 0; i < sz; i++)
            {
                TreeNode *temp = q.front();
                q.pop();

                curr.push_back(temp->val);
                if (temp->left)
                    q.push(temp->left);
                if (temp->right)
                    q.push(temp->right);
            }

            if (leftToRight)
                ans.push_back(curr);
            else
            {
                reverse(curr.begin(), curr.end());
                ans.push_back(curr);
            }

            leftToRight = !leftToRight;
        }

        return ans;
    }
};